proposals/testing/NNNN-targeted-interoperability-swift-testing-and-xctest.md

```diff
@@ -0,0 +1,292 @@
+
+## Introduction
+
+Many projects want to migrate from XCTest to Swift Testing, and may be in an
+intermediate state where test helpers written using XCTest API are called from
+Swift Testing. Today, the Swift Testing and XCTest libraries stand mostly
+independently, which means an `XCTAssert` failure in a Swift Testing test is
+silently ignored. To address this, we formally declare a set of interoperability
+principles and propose updates to specific APIs that will enable users to
+migrate with confidence.
+
+## Motivation
+
+Unfortunately, mixing an API call from one framework with a test from the other
+framework may not work as expected. As a more concrete example, if you take an
+existing test helper function written for XCTest and call it in a Swift Testing
+test, it won't report the assertion failure:
+
+```swift
+func assertUnique(_ elements: [Int]) {
+  if Set(elements).count != elements.count {
+    XCTFail("\(elements) has non unique elements")
+  }
+}
+
+// XCTest
+
+class FooTests: XCTestCase {
+    func testDups() {
+        assertUnique([1, 2, 1]) // Fails as expected
+    }
+}
+
+// Swift Testing
+
+@Test func `Duplicate elements`() {
+  assertUnique([1, 2, 1]) // Passes? Oh no!
+}
+```
+
+Generally, we get into trouble today when ALL the following conditions are met:
+
+- You call XCTest API in a Swift Testing test, or call Swift Testing API in a
+  XCTest test
+- The API doesn't function as expected in some or all cases, and
+- You get no notice at build time or runtime about the malfunction
+
+For the remainder of this proposal, we’ll describe tests which exhibit this
+problem as **lossy without interop**.
+
+If you've switched completely to Swift Testing and don't expect to use XCTest in
+the future, this proposal includes a mechanism to **prevent you from
+inadvertently introducing XCTest APIs to your project**, including via a testing
+library.
+
+## Proposed solution
+
+- **XCTest APIs which are lossy without interop will work as expected when
+  called in Swift Testing.** In addition, runtime diagnostics will be included
+  to highlight opportunities to adopt newer Swift Testing equivalent APIs.
+
+- Conversely, **Swift Testing API called in XCTest will work as expected if
+  XCTest already provides similar functionality.** You should always feel
+  empowered to choose Swift Testing when writing new tests or test helpers, as
+  it will work properly in both types of tests.
+
+We don't propose supporting interoperability for APIs without risk of data loss,
+because they naturally have high visibility. For example, using `throw XCTSkip`
+in a Swift Testing test results in a test failure rather than a test skip,
+providing a clear indication that migration is needed.
+
+## Detailed design
+
+### Highlight and support XCTest APIs which are lossy without interop
+
+We propose supporting the following XCTest APIs in Swift Testing:
+
+- Assertions: `XCTAssert*` and [unconditional failure][] `XCTFail`
```

Comment:
I realize there's not a single page in the XCTest documentation which collects all `XCTAssert` functions, but I think a close approximation of that is the [Equality and Inequality Assertions](https://developer.apple.com/documentation/xctest/equality-and-inequality-assertions) page, so you could link to that as a representative example.